rust/parser/mod.rs

```diff
@@ -49,279 +46,357 @@ use crate::{
         TypeQLUndefine, TypeQLUpdate,
     },
 };
-use antlr_rust::{
-    token::Token,
-    tree::{ParseTree, TerminalNode as ANTLRTerminalNode},
-};
 use chrono::{NaiveDate, NaiveDateTime};
-use std::rc::Rc;
+use pest::Parser;
+use pest_derive::Parser;
 
-// keep star import to not expose generated code
-use typeql_grammar::typeqlrustparser::*;
+#[derive(Parser)]
+#[grammar = "parser/typeql.pest"]
+struct TypeQLParser;
 
-type TerminalNode<'a> = ANTLRTerminalNode<'a, TypeQLRustParserContextType>;
+type SyntaxTree<'a> = pest::iterators::Pair<'a, Rule>;
+type Children<'a> = pest::iterators::Pairs<'a, Rule>;
 
-fn get_string(string: Rc<TerminalNode>) -> String {
-    unquote(&string.get_text())
+trait TreeChildren {
+    type Items;
+    fn into_children(self) -> Self::Items;
 }
 
-fn get_regex(string: Rc<TerminalNode>) -> String {
-    unescape_regex(&unquote(&string.get_text()))
+impl<'a> TreeChildren for SyntaxTree<'a> {
+    type Items = Children<'a>;
+    fn into_children(self) -> Self::Items {
+        self.into_inner()
+    }
 }
 
-fn get_long(long: Rc<TerminalNode>) -> i64 {
-    long.get_text().parse().expect(&ILLEGAL_GRAMMAR.format(&[long.get_text().as_str()]).message)
+trait RuleIterator {
+    type Item;
+
+    fn omit(&mut self, rule: Rule) -> &mut Self;
+
+    fn consume(&mut self, rule: Rule) -> Self::Item;
+    fn try_consume(&mut self, rule: Rule) -> Option<Self::Item>;
+
+    fn consume_any(&mut self) -> Self::Item;
+    fn try_consume_any(&mut self) -> Option<Self::Item>;
 }
 
-fn get_double(double: Rc<TerminalNode>) -> f64 {
-    double.get_text().parse().expect(&ILLEGAL_GRAMMAR.format(&[double.get_text().as_str()]).message)
+impl<'a, T: Iterator<Item = SyntaxTree<'a>>> RuleIterator for T {
+    type Item = SyntaxTree<'a>;
+
+    fn omit(&mut self, rule: Rule) -> &mut Self {
```

Comment:
`skip_expected`, `consume_expected`, `consume_if_matches`, `consume_any`, `try_consume_any` for clarity